### PR TITLE
ENH: Munchausen Reinforcement Learning

### PR DESCRIPTION
**Pull request by AntoineTheb**
https://en.wikipedia.org/wiki/Baron_Munchausen

### CODE REVIEW

File: article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+$$L(\phi) = \mathbb{E}_{s,a,s',r \sim D}\big(\underbrace{r + \gamma \max_{a'}Q_{\phi'}(s', \hat{a}') - \tau \log \pi_\theta(\hat{a}'|s')}_{\text{target}} - Q_\phi(s,a)\big)^2, \hat{a}' \sim \pi_\theta(\cdot||s')$$
+
+**However**, the authors propose not to subtract the log-policy to all rewards, but instead to "optimize for the immediate reward _augmented_ by the log-policy". Their motivations are as follow:
+
+If the optimal Q-function $$Q^*(s,a)$$ was known, the agent would use it as a target in equation $$(1)$$. However, it is unknown and the agent instead uses it's own Q-function as an estimate. This is bootstrapping. If the optimal policy $$\pi^*$$ was known, its log-probability for the optimal action would be $$0$$, and $$-\infty$$ otherwise. Because we don't have the optimal policy, we can instead use the log-probabilities from our own policy to add to the reward signal.
+
+# Method:
+
+Therefore, M-RL is as simple as replacing $$ r_t $$ with $$ r_t + \alpha \log \pi(a_t{\mid}s_t)$$ anytime TD is involved. Because M-RL supposes a stochastic policy and DQN typically trains implicit deterministic policies, the authors suggest two changes to the typical DQN update:
```
Comment (jodoin) [inline article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md:44]: update.

**Review by jodoin — APPROVED** [review]

File: article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
+---
+layout: review
+title: "Munchausen Reinforcement Learning"
+tags: reinforcement
+author: "Antoine Théberge"
+cite:
+    authors: "Nino Vieillard, Olivier Pietquin, Matthieu Geist"
+    title:   "Munchausen Reinforcement Learning"
+    venue:   "https://arxiv.org/abs/2007.14430"
+pdf: ""
```
Comment (jhlegarreta) [inline article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md:10]: Should the `venue` be `arXiv` and the `pdf` field contain the link to the PDF?

File: article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+> source: [https://en.wikipedia.org/wiki/M%C3%BCnchhausen_trilemma](https://en.wikipedia.org/wiki/M%C3%BCnchhausen_trilemma)
+
+# Introduction
+
+Reinforcement learning (RL) often uses _bootstrapping_ when training agents: bootstrapping occurs when the target for computing a loss function leverages the network's own estimates, like with Deep Q-Network (DQN)'s Temporal-Difference (TD) update: 
+
+$$\tag{1} L(\phi) = \mathbb{E}_{s,a,s',r \sim D}\big(\underbrace{r + \gamma \max_{a'}Q_{\phi'}(s', a')}_{\text{target}} - Q_\phi(s,a)\big)^2$$
+
+The authors propose to leverage another estimate from the agent: it's own policy. It is similar to maximum-entropy RL, which attempts to maximize both the expected return and the policy's entropy s.t. the RL problem becomes:
```
Comment (jhlegarreta) [inline article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md:30]: Not sure if the last sentence is missing something: I feel that the s.t. is missing its "argument", something like
_It is similar to maximum-entropy RL, which attempts to maximize both the expected return and the policy's entropy s.t. {missing}, the RL problem becoming:_

File: article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+$$\pi^* = \argmax_\pi \mathbb{E}\Big[\sum_{t=0}^\infty r_t + \alpha \mathbb{H}(\pi(\cdot||s_t))\Big],$$
+
+leading to the (approximate)[^1] following loss function for the Q-network:
+
+$$L(\phi) = \mathbb{E}_{s,a,s',r \sim D}\big(\underbrace{r + \gamma \max_{a'}Q_{\phi'}(s', \hat{a}') - \tau \log \pi_\theta(\hat{a}'|s')}_{\text{target}} - Q_\phi(s,a)\big)^2, \hat{a}' \sim \pi_\theta(\cdot||s')$$
+
+**However**, the authors propose not to subtract the log-policy to all rewards, but instead to "optimize for the immediate reward _augmented_ by the log-policy". Their motivations are as follow:
```
Comment (jhlegarreta) [inline article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md:38]: as follows?

File: article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+If the optimal Q-function $$Q^*(s,a)$$ was known, the agent would use it as a target in equation $$(1)$$. However, it is unknown and the agent instead uses it's own Q-function as an estimate. This is bootstrapping. If the optimal policy $$\pi^*$$ was known, its log-probability for the optimal action would be $$0$$, and $$-\infty$$ otherwise. Because we don't have the optimal policy, we can instead use the log-probabilities from our own policy to add to the reward signal.
```
Comment (jhlegarreta) [inline article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md:40]: _it's own_ : _its own_

**Review by jhlegarreta — COMMENTED** [review]
A few in-line comments.

File: article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md (hunk 3 above)
Comment (AntoineTheb) [reply article/_posts/2020-09-10-Munchausen_Reinforcement_Learning.md]: I reformulated it in b0c97df

**Review by jhlegarreta — APPROVED** [review]

**Pull request merged**